aws/dx_vif.go

```diff
@@ -0,0 +1,256 @@
+		Pending:    pending,
+		Target:     target,
+		Refresh:    dxVirtualInterfaceStateRefresh(conn, d.Id()),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      10 * time.Second,
+		MinTimeout: 5 * time.Second,
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for Direct Connect virtual interface %s to become available: %s", d.Id(), err.Error())
+	}
+
+	return nil
+}
+
+// Attributes common to public VIFs and creator side of hosted public VIFs.
+func dxPublicVirtualInterfaceAttributes(d *schema.ResourceData, meta interface{}, vif *directconnect.VirtualInterface) error {
+	if err := dxVirtualInterfaceAttributes(d, meta, vif); err != nil {
+		return err
+	}
+	d.Set("route_filter_prefixes", flattenDxRouteFilterPrefixes(vif.RouteFilterPrefixes))
+
+	return nil
+}
+
+// Attributes common to private VIFs and creator side of hosted private VIFs.
+func dxPrivateVirtualInterfaceAttributes(d *schema.ResourceData, meta interface{}, vif *directconnect.VirtualInterface) error {
+	return dxVirtualInterfaceAttributes(d, meta, vif)
+}
+
+// Attributes common to public/private VIFs and creator side of hosted public/private VIFs.
+func dxVirtualInterfaceAttributes(d *schema.ResourceData, meta interface{}, vif *directconnect.VirtualInterface) error {
+	if err := dxVirtualInterfaceArnAttribute(d, meta); err != nil {
+		return err
+	}
+
+	d.Set("connection_id", vif.ConnectionId)
+	d.Set("name", vif.VirtualInterfaceName)
+	d.Set("vlan", vif.Vlan)
+	d.Set("bgp_asn", vif.Asn)
+	d.Set("bgp_auth_key", vif.AuthKey)
+	d.Set("address_family", vif.AddressFamily)
+	d.Set("customer_address", vif.CustomerAddress)
+	d.Set("amazon_address", vif.AmazonAddress)
+
+	return nil
+}
+
+func dxVirtualInterfaceArnAttribute(d *schema.ResourceData, meta interface{}) error {
+	arn := arn.ARN{
+		Partition: meta.(*AWSClient).partition,
+		Region:    meta.(*AWSClient).region,
+		Service:   "directconnect",
+		AccountID: meta.(*AWSClient).accountid,
+		Resource:  fmt.Sprintf("dxvif/%s", d.Id()),
+	}.String()
+	d.Set("arn", arn)
+
+	return nil
+}
+
+func expandDxRouteFilterPrefixes(cfg []interface{}) []*directconnect.RouteFilterPrefix {
+	prefixes := make([]*directconnect.RouteFilterPrefix, len(cfg), len(cfg))
+	for i, p := range cfg {
+		prefix := &directconnect.RouteFilterPrefix{
+			Cidr: aws.String(p.(string)),
+		}
+		prefixes[i] = prefix
+	}
+	return prefixes
+}
+
+func flattenDxRouteFilterPrefixes(prefixes []*directconnect.RouteFilterPrefix) *schema.Set {
+	out := make([]interface{}, 0)
+	for _, prefix := range prefixes {
+		out = append(out, aws.StringValue(prefix.Cidr))
+	}
+	return schema.NewSet(schema.HashString, out)
+}
+
+// Schemas common to all (public/private, hosted or not) virtual interfaces.
+var dxVirtualInterfaceSchemaWithTags = mergeSchemas(
+	dxVirtualInterfaceSchema,
+	map[string]*schema.Schema{
+		"tags": tagsSchema(),
+	},
+)
+var dxVirtualInterfaceSchema = map[string]*schema.Schema{
```

Comment:
Do you mind moving all the schema related code to the top of the file? It's just a convention we tend to follow.

Reply:
Will do.

Reply:
Done.